cmd/refresh-uids-from-ferry/utils.go

```diff
@@ -253,16 +254,23 @@ func checkFerryDataInDB(ferryData, dbData []db.FerryUIDDatum) bool {
 // authFunc.  It spins up a worker to get data from FERRY, and then puts that data into
 // a channel for aggregation.
 func getAndAggregateFERRYData(ctx context.Context, username string, authFunc func() func(context.Context, string, string) (*http.Response, error),
```

Comment:
This is interesting, so it's a function that returns another anonymous function? Does this syntax mean it returns this: (*http.Response, error)?

Reply:
In Go, functions are first-class values, so you can pass them into functions and return them.  So you'll see a LOT of decorator-pattern-type stuff in Go.

So yes, this is a function that takes among other things a function that itself returns a function that takes a context and two strings and returns (*http.Response, error)`.  You can think of it like (this is NOT valid Go syntax, but just for illustrative purposes):

`authFunc func() (func(context.Context, string, string) (*http.Response, error))`

The simplest way you'd use this (ignoring all logic) is something like:

```go

myAuthFunc := func() func(context.Context, string, string) (*http.Response, error) {
    return func(ctx context.Context, a string, b string) (*http.Response, error) {
       check_something_with_context(ctx)
       maybe_do_something_with_a_and_b(a, b)
      return &http.Response{}, nil
   }
}

err := getAndAggregateFERRYData(ctx, "cathulhu", myAuthFunc, otherVariables...)
```

In this code, the two functions in [ferryAuthMethods.go](https://github.com/fermitools/managed-tokens/blob/issue-98/cmd/refresh-uids-from-ferry/ferryAuthMethods.go) are used as the `authFunc` arguments.  The authFunc argument is this type (a function that returns a function) to allow for different authorization function "decorators" to all conform to a standard when they're used - that is, the return value of those auth functions must be a function that uses a context and two strings and returns the response and an error.